### PR TITLE
(WIP) Update error handling examples

### PR DESCRIPTION
**Pull request by danicodeve**
@royanger 

The error handling examples are out-of-date. Updated error handling React examples with a simple form and display of error messages.

Other items considered but did not add to maintain simplicity:
- import ClerkAPIError from '@clerk/types' for more strongly typed examples
- aria attributes such as aria-invalid and aria-errormessage
- autocomplete and placeholder attributes of input elements

### CODE REVIEW

File: docs/custom-flows/error-handling.mdx
Comment (JacobMGEvans) [inline docs/custom-flows/error-handling.mdx:30]: Can you elaborate on what this is doing. We just need to be clear for documentation.
Likely, clearing a previously set error from a form submission attempt?

File: docs/custom-flows/error-handling.mdx
Comment (JacobMGEvans) [inline docs/custom-flows/error-handling.mdx:22]:
```suggestion
  const [error, setError] = useState<Error>();
```

File: docs/custom-flows/error-handling.mdx
Comment (JacobMGEvans) [inline docs/custom-flows/error-handling.mdx:44]: ignore my lack of formatting 😅 
```suggestion
    } catch(err: unknown) {
    if(err instanceof Error) setError(err);
    return; 
    }
```

File: docs/custom-flows/error-handling.mdx
Comment (JacobMGEvans) [inline docs/custom-flows/error-handling.mdx:35]: Instead of optional chaining, check `signUp` is ready in scope.

File: docs/custom-flows/error-handling.mdx
Comment (JacobMGEvans) [inline docs/custom-flows/error-handling.mdx:32]:
```suggestion
    if (!isLoaded)  return; 
```

**Review by JacobMGEvans — CHANGES_REQUESTED** [review]
Changes suggested in first example apply to both. Appreciate the PR and happy to help you get this in 😄

File: docs/custom-flows/error-handling.mdx
Comment (danicodeve) [reply docs/custom-flows/error-handling.mdx]: I've implemented this change, and it works. But note that TypeScript will complain when accessing properties that are not part of the standard built-in Error object. The error returned by Clerk has extra properties like 'clerkError', 'clerkTraceId', and 'errors'.

Comment (JacobMGEvans) [reply docs/custom-flows/error-handling.mdx]: I wont lie, I forgot that existed

Comment (danicodeve) [reply docs/custom-flows/error-handling.mdx]: Implemented error type checking.

Used `isClerkAPIResponseError` to do type checking on the actual response. This it to avoid changing type of err to any. Details as to why this is preferable here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#unknown-on-catch-clause-bindings

**Review by JacobMGEvans — APPROVED** [review]
Tentative to @royanger's approval

**Comment by royanger** [comment]
Let me take a good look at this later today. Just handling a bunch of v5 stuff rightn ow.

**Comment by royanger** [comment]
This will be blocked on V5 -- error types are changing a little bit. I'll review this week and update for the new type.

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:40]: Let's add error handling in here as well, using the same logic as line 44

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:17]:
```suggestion
import { useSignUp } from "@clerk/clerk-react";
import { isClerkAPIResponseError } from "@clerk/clerk-react/errors";
```
Errors will be from the /errors subpath in v5

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:27]:
```suggestion
  const attemptSignUp = async (e: React.FormEvent) => {
```
This is consistent with updated docs.

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:16]:
```suggestion
import { useState } from "react";
```
See change below.

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:99]:
```suggestion
import { useState } from "react";
```

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:100]:
```suggestion
import { useSignIn } from "@clerk/clerk-react";
import { isClerkAPIResponseError } from "@clerk/clerk-react/errors";
```

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:110]:
```suggestion
  const attemptSignIn = async (e: React.FormEvent) => {
```

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [inline docs/custom-flows/error-handling.mdx:123]: Let's add error handling here as well

**Review by royanger — CHANGES_REQUESTED** [review]

**Comment by royanger** [comment]
Looks good. I requested a few minor updates/changes, mostly for consistency with other docs.

Can you change this to be against branch release/v5?

Thanks so much for the submission!

File: docs/custom-flows/error-handling.mdx
Comment (danicodeve) [reply docs/custom-flows/error-handling.mdx:41]: Any error thrown by `setActive` is already caught by the existing catch clause.

Could you clarify what extra error handling is needed? Are errors thrown by setActive of a different type, such as ClerkRuntimeError, and thus need to be handled differently?

**Comment by danicodeve** [comment]
Thanks for the review and comments @royanger . 

I've made the requested changes, except for one where I've asked for clarification.

File: docs/custom-flows/error-handling.mdx
Comment (royanger) [reply docs/custom-flows/error-handling.mdx:41]: The catch will handles an error, but not `result.status === 'missing_requirements' | 'abandoned'`. See https://clerk.com/docs/references/react/use-sign-up#result-status

This is a non-error return that still should be read. The `abandoned` is something users can run into if they take too long to try and confirm signUp, depending on signUp method. `missing_requirements` will *usually* be found and resolved during dev, but in case the app setting change its good to show versus a silent fail with no message.

**Comment by royanger** [comment]
> Thanks for the review and comments @royanger .
> 
> I've made the requested changes, except for one where I've asked for clarification.

Thank you! I added info there. It was something that caught me off guard when working with the flows early on. We have errors, and successful returns...that are errors.

**Comment by royanger** [comment]
@alexisintech -- we're just finishing up this PR with a couple of last changes, but it will be ready for your review soon. This is ready for v5 with the (small) error import change, so it against the `release/v5` branch.

File: docs/custom-flows/error-handling.mdx
Comment (danicodeve) [reply docs/custom-flows/error-handling.mdx:41]: Got it. These successful errors do not conform to the type `ClerkAPIError` and cannot be handled immediately by the same logic as line 44. Here are two approaches:

1. Create an object that conforms to ClerkAPIError, set the error and it is shown in the UI.
```typescript
if (result.status == 'missing_requirements') {
        const error = { code: result.status, message: 'Requirement from the email, phone, username settings is missing.' }
        setErrors([error]);
        return;
}
```

2. Throw a standard Error object and show it in the console.
```typescript
    if (result.status == 'missing_requirements') {
        throw new Error(result.status);
        return;
    }
} catch(err: unknown) {
    if (isClerkAPIResponseError(err)) {
        setErrors(err.errors);
        return;
    }
    if (err instanceof Error) { 
        console.error(err.message) 
        return;
    }
}
```

Do you have a preference between these two, or is there another approach you'd prefer?

Comment (royanger) [reply docs/custom-flows/error-handling.mdx:41]: Definitely 1 or something like it. The goal with this is two fold. 1. Show the kinds of errors that need to be dealt with and 2. Give an example of showing those to the user.

If a dev wants to handle these different they can definitely can -- this gives them the info what what to handle.

**Comment by royanger** [comment]
Just copying this from the conversation (that I am resolving) to here, so @alexisintech can see and answer.

If Alexis isn't working on this, a PR to update would be awesome.

From @danicodeve:
PR is now updated to show errors that need to be dealt, and a way to show them to the user.

Related:
SignIn result status values on docs [React/useSignIn](https://github.com/clerk/clerk-docs/blob/release/v5/docs/references/react/use-sign-in.mdx) and [JS/SignIn](https://github.com/clerk/clerk-docs/blob/release/v5/docs/references/javascript/sign-in/sign-in.mdx) (current and v5) are inconsistent with those from SignInStatus type in @clerk/types. Unsure if @alexisintech is already working on this. If needed, I'm happy to create another PR for this separate issue.

SignUp result status values are consistent with @clerk/types. 👍

**Comment by royanger** [comment]
This looks good to me! Thanks for this. @alexisintech can review, but code wise this is a great update.

**Review by royanger — APPROVED** [review]

**Comment by alexisintech** [comment]
> Just copying this from the conversation (that I am resolving) to here, so @alexisintech can see and answer.
> 
> If Alexis isn't working on this, a PR to update would be awesome.
> 
> From @danicodeve: PR is now updated to show errors that need to be dealt, and a way to show them to the user.
> 
> Related: SignIn result status values on docs [React/useSignIn](https://github.com/clerk/clerk-docs/blob/release/v5/docs/references/react/use-sign-in.mdx) and [JS/SignIn](https://github.com/clerk/clerk-docs/blob/release/v5/docs/references/javascript/sign-in/sign-in.mdx) (current and v5) are inconsistent with those from SignInStatus type in @clerk/types. Unsure if @alexisintech is already working on this. If needed, I'm happy to create another PR for this separate issue.
> 
> SignUp result status values are consistent with @clerk/types. 👍

I updated the react/useSignIn and JS/SignIn pages recently ([PR](https://github.com/clerk/clerk-docs/pull/681)) - is the information correct, or are the inconsistencies with those from SignInStatus type in @clerk/types still present? If still present, can you point out what's inconsistent so I can update the docs accordingly?

**Comment by royanger** [comment]
> I updated the react/useSignIn and JS/SignIn pages recently ([PR](https://github.com/clerk/clerk-docs/pull/681)) - is the information correct, or are the inconsistencies with those from SignInStatus type in @clerk/types still present? If still present, can you point out what's inconsistent so I can update the docs accordingly?

I haven't checked the new docs, but here is the Type:

https://github.com/clerk/javascript/blob/92b0d54d7a069fa168a74078654c55e6eb9bdbde/packages/types/src/signIn.ts#L98

Note: I am trying to confirm that `needs_new_password` should be removed.

**Comment by alexisintech** [comment]
> JS/SignIn

okay yes, so I did check that type when I updated the docs, and the docs are accurate now! :)

**Comment by alexisintech** [comment]
> > I updated the react/useSignIn and JS/SignIn pages recently ([PR](https://github.com/clerk/clerk-docs/pull/681)) - is the information correct, or are the inconsistencies with those from SignInStatus type in @clerk/types still present? If still present, can you point out what's inconsistent so I can update the docs accordingly?
> 
> I haven't checked the new docs, but here is the Type:
> 
> https://github.com/clerk/javascript/blob/92b0d54d7a069fa168a74078654c55e6eb9bdbde/packages/types/src/signIn.ts#L98
> 
> Note: I am trying to confirm that `needs_new_password` should be removed.

okay yes, so I did check that type when I updated the docs, and the docs are accurate now! :)

**Comment by alexisintech** [comment]
@royanger if you tested the code and you approve, then I approve! I will let you merge just in case anything else needs to be added

**Review by alexisintech — APPROVED** [review]
wow! awesome work. thank you so much for the contribution 😸💖

**Comment by royanger** [comment]
> @royanger if you tested the code and you approve, then I approve! I will let you merge just in case anything else needs to be added

Let me wait to hear back about `needs_new_password` and I will merge after.

**Pull request merged**
